clang/tools/driver/cc1_main.cpp

```diff
@@ -283,6 +285,94 @@ class ObjectStoreCachingOutputs : public CachingOutputs {
   Optional<llvm::cas::ObjectRef> DependenciesOutput;
 };
 
+/// An \p OutputBackend that just records the list of output paths/names.
+class CollectingOutputBackend : public llvm::vfs::ProxyOutputBackend {
```

Comment:
How hard would it be to split all the caching backend code out of cc1_main.cpp? Ideally we could have one .cpp for each caching backend plus one for the common code. Upstream this file is only ~300 lines, but here it is now ~1200.  It seems bad to have cc1_main be 75% caching-related.

Reply:
Good idea, this seems straightforward by pulling the classes out of `cc1_main.cpp`.
But I'd like to do that as follow-up after eliminating any internal merge conflict.

Reply:
sgtm